File: cplus/leetcode/202502/algo1473.cpp
#include <bits/stdc++.h>
using namespace std;

// 动态规划 - 约束划分个数
class Solution {
public:
    int minCost(vector<int>& houses, vector<vector<int>>& cost, int m, int n, int target) {
        // 自行解答
        vector<vector<vector<int>>> memo(m,vector<vector<int>>(n+1,vector<int>(target+1, -1)));

        auto dfs = [&] (this auto&& dfs, int i, int p, int l) ->int {
            if (l < 0) return INT_MAX/2;
            if (i < 0)
                if (l == 0) return 0;
                else return INT_MAX/2;      // 使用 INT_MAX 做加法越界

            auto& res = memo[i][p][l];
            if (res != -1) return res;
            
            res = INT_MAX/2;
            if (houses[i]!=0) return res = dfs(i-1, houses[i], l-(houses[i]!=p));
            
            for (int c=1; c<=n; c++) {
                res = min(res, cost[i][c-1] + dfs(i-1, c, l - (c!=p)));
            }
            return res;
        };

        int ans = dfs(m-1, 0, target);
        return (ans >= INT_MAX/2) ? -1 : ans;

    }
};

File: python/algo/202502/algo1473.py
from cmath import inf
from functools import cache
from typing import List

# 动态规划 - 约束划分个数
class Solution:
    def minCost(self, houses: List[int], cost: List[List[int]], m: int, n: int, target: int) -> int:
        m = len(houses)

        @cache
        def dfs(i:int, prev:int, left:int) -> int:
            if left < 0: return inf
            if i < 0:
                if left == 0: return 0
                else: return inf

            if (houses[i]!=0): return dfs(i-1, houses[i], left-(prev!=houses[i]))
            
            res = inf
            for c in range(1, n+1):
                res = min(res, cost[i][c-1] + dfs(i-1, c, left-(prev!=c)))
            return res
        
        ans = dfs(m-1, 0, target)
        return -1 if ans == inf else ans
            